crates/iceberg/src/spec/values.rs

```diff
@@ -0,0 +1,417 @@
+    /// 16-byte big-endian value
+    UUID(i128),
+    /// Binary value
+    Fixed(usize, Vec<u8>),
+    /// Binary value (without length)
+    Binary(Vec<u8>),
+    /// Stores unscaled value as two’s-complement big-endian binary,
+    /// using the minimum number of bytes for the value
+    Decimal(Decimal),
+    /// A struct is a tuple of typed values. Each field in the tuple is named and has an integer id that is unique in the table schema.
+    /// Each field can be either optional or required, meaning that values can (or cannot) be null. Fields may be any type.
+    /// Fields may have an optional comment or doc string. Fields can have default values.
+    Struct(Struct),
+    /// A list is a collection of values with some element type.
+    /// The element field has an integer id that is unique in the table schema.
+    /// Elements can be either optional or required. Element types may be any type.
+    List(Vec<Option<Value>>),
+    /// A map is a collection of key-value pairs with a key type and a value type.
+    /// Both the key field and value field each have an integer id that is unique in the table schema.
+    /// Map keys are required and map values can be either optional or required. Both map keys and map values may be any type, including nested types.
+    Map(HashMap<String, Option<Value>>),
+}
+
+impl From<Value> for ByteBuf {
+    fn from(value: Value) -> Self {
+        match value {
+            Value::Boolean(val) => {
+                if val {
+                    ByteBuf::from([0u8])
+                } else {
+                    ByteBuf::from([1u8])
+                }
+            }
+            Value::Int(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::LongInt(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Float(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Double(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Date(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Time(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::Timestamp(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::TimestampTZ(val) => ByteBuf::from(val.to_le_bytes()),
+            Value::String(val) => ByteBuf::from(val.as_bytes()),
+            Value::UUID(val) => ByteBuf::from(val.to_be_bytes()),
+            Value::Fixed(_, val) => ByteBuf::from(val),
+            Value::Binary(val) => ByteBuf::from(val),
+            _ => todo!(),
+        }
+    }
+}
+
+/// The partition struct stores the tuple of partition values for each file.
+/// Its type is derived from the partition fields of the partition spec used to write the manifest file.
+/// In v2, the partition struct’s field ids must match the ids from the partition spec.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Struct {
+    /// Vector to store the field values
+    fields: Vec<Option<Value>>,
+    /// A lookup that matches the field name to the entry in the vector
+    lookup: HashMap<String, usize>,
```

Comment:
Use lookup will make the memory cost if we have multiple struct with same type.

One solve way is to use `Arc<StructType>` in Struct. I try this design in https://github.com/icelake-io/icelake/pull/136. 

Anyway I think it's ok to solve this Problem in another PR. If this desgin looks work, I'm glad to port it.

Reply:
I tried using `Arc<Structype>` but I think for that `NestedField` has to be changed to `field_type: Arc<Type>` because if I have a struct that is part of a nested struct, I need to be able to reference the type(StructType) of the inner struct. I would like to split that change into another PR and that's why I used your first implementation for now.

Reply:
Implement it in another PR sounds reasonable for me. And does this problem can solve by make Struct be Arc rather than Type, like:
```
struct Type {
    ...
    Struct(Arc<Struct>)
    ...
}
```

The benefit of this is that we can use `Type` directly.

Reply:
We can try it for the follow up PR.